### PR TITLE
Pose graph: add pose modifying nodes

### PR DESCRIPTION
**Pull request by shrinktofit**
Re: #

### Changelog

* This PR adds the following pose modifying nodes:
  - `ApplyTransform` apply a transform, in position term and rotation term, to an animated node of a pose.
  - `CopyTransform` copies transform of one animated node to another animated node of a pose.
  - `TwoBoneIKSolver` solves two bone ik problem.

-------

### Continuous Integration

This pull request:

* [ ] needs automatic test cases check.
  > Manual trigger with `@cocos-robot run test cases` afterward.
* [ ] does not change any runtime related code or build configuration
  > If any reviewer thinks the CI checks are needed, please uncheck this option, then close and reopen the issue.

-------

### Compatibility Check

This pull request:

* [ ] changes public API, and have ensured backward compatibility with [deprecated features](https://github.com/cocos/cocos-engine/blob/v3.5.0/docs/contribution/deprecated-features.md).
* [ ] affects platform compatibility, e.g. system version, browser version, platform sdk version, platform toolchain, language version, hardware compatibility etc.
* [ ] affects file structure of the build package or build configuration which requires user project upgrade.
* [ ] **introduces breaking changes**, please list all changes, affected features and the scope of violation.

### CODE REVIEW

File: cocos/animation/marionette/pose-graph/pose-nodes/apply-transform.ts [hunk 1]
```diff
@@ -0,0 +1,266 @@
...
+
+    @serializable
+    @editable
+    @type(TransformOperation)
+    public positionOperation = TransformOperation.LEAVE_UNCHANGED;
+
+    @serializable
+    @editable
+    @input({ type: PoseGraphType.VEC3 })
+    @visible(function (this: PoseNodeApplyTransform) { return this.positionOperation !== TransformOperation.LEAVE_UNCHANGED; })
```
Comment (Canvasfull) [inline cocos/animation/marionette/pose-graph/pose-nodes/apply-transform.ts]: ESLint 检测 unnamed function.

File: cocos/animation/marionette/pose-graph/pose-nodes/apply-transform.ts [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+
+    @serializable
+    @editable
+    @type(TransformOperation)
+    public rotationOperation = TransformOperation.LEAVE_UNCHANGED;
+
+    @serializable
+    @editable
+    @input({ type: PoseGraphType.QUAT })
+    @visible(function (this: PoseNodeApplyTransform) { return this.rotationOperation !== TransformOperation.LEAVE_UNCHANGED; })
```
Comment (Canvasfull) [inline cocos/animation/marionette/pose-graph/pose-nodes/apply-transform.ts]: ESLint 检测 unnamed function.

File: cocos/animation/marionette/pose-graph/pose-nodes/ik/two-bone-ik-solver.ts [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+            endEffectorTargetPosition,
+            poleTargetPosition,
+            this.debug ? this : undefined,
+        );
+
+        modificationQueue.push(iRootTransform, rootTransform);
+        modificationQueue.push(iMiddleTransform, middleTransform);
+        modificationQueue.push(iEndEffectorTransform, endEffectorTransform);
+
+        // // TODO: bad performance!
```
Comment (Canvasfull) [inline cocos/animation/marionette/pose-graph/pose-nodes/ik/two-bone-ik-solver.ts]: 以下注视掉的代码需要保留吗？有什么用？

File: cocos/animation/marionette/pose-graph/pose-nodes/ik/two-bone-ik-solver.ts [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+    }
+
+    @serializable
+    @editable
+    @type(TargetSpecificationType)
+    public type = TargetSpecificationType.VALUE;
+
+    @serializable
+    @editable
+    @visible(function (this: TargetSpecification) { return this.type === TargetSpecificationType.VALUE; })
```
Comment (Canvasfull) [inline cocos/animation/marionette/pose-graph/pose-nodes/ik/two-bone-ik-solver.ts]: ESlint unnamed function.

File: cocos/animation/marionette/pose-graph/pose-nodes/ik/two-bone-ik-solver.ts [hunk 5]
```diff
@@ -0,0 +1,260 @@
...
+    public targetPosition = new Vec3();
+
+    @serializable
+    @editable
+    @type(TransformSpace)
+    @visible(function (this: TargetSpecification) { return this.type === TargetSpecificationType.VALUE; })
```
Comment (Canvasfull) [inline cocos/animation/marionette/pose-graph/pose-nodes/ik/two-bone-ik-solver.ts]: ESlint unnamed function.

File: cocos/animation/marionette/pose-graph/pose-nodes/ik/two-bone-ik-solver.ts [hunk 6]
```diff
@@ -0,0 +1,260 @@
...
+    public targetPositionSpace = TransformSpace.WORLD;
+
+    @serializable
+    @editable
+    @visible(function (this: TargetSpecification) { return this.type === TargetSpecificationType.BONE; })
```
Comment (Canvasfull) [inline cocos/animation/marionette/pose-graph/pose-nodes/ik/two-bone-ik-solver.ts]: ESlint unnamed function.

File: cocos/animation/marionette/pose-graph/pose-nodes/ik/two-bone-ik-solver.ts [hunk 7]
```diff
@@ -0,0 +1,260 @@
...
+
+        const {
+            hRoot: { index: iRootTransform },
+            hMiddle: { index: iMiddleTransform },
+            hEndEffector: { index: iEndEffectorTransform },
+        } = workspace;
+
+        // // TODO: bad performance!
+        // // Save local space pose.
+        // const localPose = context.pushDuplicatedPose(inputPose);
```
Comment (Canvasfull) [inline cocos/animation/marionette/pose-graph/pose-nodes/ik/two-bone-ik-solver.ts]: 已经注释的代码 可以删除。这里有什么用处？

File: cocos/animation/marionette/pose-graph/pose-nodes/ik/two-bone-ik-solver.ts [hunk 8]
```diff
@@ -0,0 +1,260 @@
...
+            this.debug ? this : undefined,
+        );
+
+        modificationQueue.push(iRootTransform, rootTransform);
+        modificationQueue.push(iMiddleTransform, middleTransform);
+        modificationQueue.push(iEndEffectorTransform, endEffectorTransform);
+
+        // // TODO: bad performance!
+        // {
+        //     // Push transforms.
```
Comment (Canvasfull) [inline cocos/animation/marionette/pose-graph/pose-nodes/ik/two-bone-ik-solver.ts]: 以下注视的代码可以考虑移除。

Comment (shrinktofit) [reply cocos/animation/marionette/pose-graph/pose-nodes/ik/two-bone-ik-solver.ts]: Solved in f4a158975ca9629e1bd7969108f186b14c3de1c4 .

File: cocos/animation/marionette/pose-graph/pose-nodes/ik/two-bone-ik-solver.ts (hunk 7 above)
Comment (shrinktofit) [reply cocos/animation/marionette/pose-graph/pose-nodes/ik/two-bone-ik-solver.ts]: Solved in f4a158975ca9629e1bd7969108f186b14c3de1c4 .

File: cocos/animation/marionette/pose-graph/pose-nodes/ik/two-bone-ik-solver.ts (hunk 6 above)
Comment (shrinktofit) [reply cocos/animation/marionette/pose-graph/pose-nodes/ik/two-bone-ik-solver.ts]: Solved in f4a158975ca9629e1bd7969108f186b14c3de1c4 .

File: cocos/animation/marionette/pose-graph/pose-nodes/ik/two-bone-ik-solver.ts (hunk 5 above)
Comment (shrinktofit) [reply cocos/animation/marionette/pose-graph/pose-nodes/ik/two-bone-ik-solver.ts]: Solved in f4a158975ca9629e1bd7969108f186b14c3de1c4 .

File: cocos/animation/marionette/pose-graph/pose-nodes/ik/two-bone-ik-solver.ts (hunk 4 above)
Comment (shrinktofit) [reply cocos/animation/marionette/pose-graph/pose-nodes/ik/two-bone-ik-solver.ts]: Solved in f4a158975ca9629e1bd7969108f186b14c3de1c4 .

File: cocos/animation/marionette/pose-graph/pose-nodes/ik/two-bone-ik-solver.ts (hunk 3 above)
Comment (shrinktofit) [reply cocos/animation/marionette/pose-graph/pose-nodes/ik/two-bone-ik-solver.ts]: Solved in f4a158975ca9629e1bd7969108f186b14c3de1c4 .

File: cocos/animation/marionette/pose-graph/pose-nodes/apply-transform.ts (hunk 2 above)
Comment (shrinktofit) [reply cocos/animation/marionette/pose-graph/pose-nodes/apply-transform.ts]: Solved in f4a158975ca9629e1bd7969108f186b14c3de1c4 .

File: cocos/animation/marionette/pose-graph/pose-nodes/apply-transform.ts (hunk 1 above)
Comment (shrinktofit) [reply cocos/animation/marionette/pose-graph/pose-nodes/apply-transform.ts]: Solved in f4a158975ca9629e1bd7969108f186b14c3de1c4 .

**Review by Canvasfull — APPROVED** [review]

**Review by Canvasfull — APPROVED** [review]

**Review by SantyWang — APPROVED** [review]

**Pull request merged**